src/stacked_borrows/mod.rs

```diff
@@ -906,26 +788,26 @@ trait EvalContextPrivExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     };
                     let item = Item::new(new_tag, perm, protected);
                     let mut global = this.machine.stacked_borrows.as_ref().unwrap().borrow_mut();
-                    let threads = &this.machine.threads;
-                    stacked_borrows.for_each(range, |offset, stack, history, exposed_tags| {
-                        stack.grant(
-                            orig_tag,
-                            item,
-                            (alloc_id, range, offset),
-                            &mut global,
-                            current_span,
-                            history,
-                            exposed_tags,
-                            threads,
-                        )
+                    let dcx = DiagnosticCxBuilder::retag(
+                        this.machine.current_span(*this.tcx),
```

Comment:
This is still creating a 2nd current_span like before. You just removed the comment that points out that this is happening...

Reply:
There are actually less than before. Hard to believe...

I think the reason this is a mess is because we now take `current_span` by value, instead of a `&mut` to one. I tried keeping it as a mutable reference, and I can't figure out the lifetime annotations that are required, if it's even possible. I pushed a commit just now that makes changes I'm pretty confident in, but clearly it doesn't work, and from this stage I don't know what to do other than do what the compiler suggests... which in this case does not get me to a building state. Help?

Reply:
Debugging other people's lifetime mess is even harder than debugging my own lifetime mess.^^

But looking at the code, it seems like already on master, the [first current_span](https://github.com/rust-lang/miri/blob/ac655ce7fe74831e865e5040d7190406143c6651/src/stacked_borrows/mod.rs#L754) is only ever used to be passed into the closure. Which means we might as well create it inside that closure. (Which is probably what you had before I complained.)  And then the lifetime trouble disappears. I have pushed that.

Reply:
Ahhh that's good insight. Thanks!